### PR TITLE
Instrument cmssw for Intel VTune

### PR DESCRIPTION
**Pull request by fwyzard**
f953447 adds a new service to CMSSW, VTuneProfilerService, that make the internal CMSSW transitions known to Intel VTune:
  - events
  - paths and modules

3d31fd9 makes signal 38 (`SIGRTMIN+4`) available for VTune

### CODE REVIEW

**Comment by fwyzard** [comment]
For this to work, a intel-vtune SCRAM tool needs to be built and added to CMSSW (see https://github.com/cms-sw/cmsdist/pull/2680 and https://github.com/cms-sw/cmsdist/pull/2681) and the CMSSW configuration needs to be updated to support it (see https://github.com/cms-sw/cmssw-config/pull/56).

**Comment by fwyzard** [comment]
A standalone test is
```
cmsrel CMSSW_9_0_X_2016-11-24-2300
cd CMSSW_9_0_X_2016-11-24-2300
cmsenv

cp ~fwyzard/public/Self.xml config/
scram setup self
cmsenv

cp ~fwyzard/public/intel-vtune.xml config/toolbox/slc6_amd64_gcc530/tools/available/intel-vtune.xml 
scram setup intel-vtune
cmsenv

git cms-init
git cms-merge-topic 16763
scram b
```

**Comment by fwyzard** [comment]
@Dr15Jones this builds on top of the comments you made last year about VTune

**Comment by fwyzard** [comment]
Then, the service can be added to any CMSSW configuration with
```
process.VTuneProfilerService = cms.Service('VTuneProfilerService',
  enableOnlyModules = cms.untracked.vstring(),
  showDelayedModules = cms.untracked.bool(True)
)
```
and VTune can be run with

```
amplxe-cl -collect general-exploration -k enable-user-tasks=true -k collect-memory-bandwidth=true -strategy=ld-linux.so.2:nt:nt,ld-linux-x86-64.so.2:nt:nt -data-limit=0 -target-duration-type=long -- cmsRun job.py
```

**Comment by fwyzard** [comment]
@rovere @felicepantaleo FYI

**Comment by fwyzard** [comment]
this should now work out of the box with CMSSW_9_0_X_2016-11-28-2300 and later

**Comment by fwyzard** [comment]
please test

**Comment by smuzaffar** [comment]
hold
we need to think about it as this will fail on machines where we do not have afs or vtune (docker containers for IBs where we do not have AFS) . I guess this will also fail for slc7 ...right?

**Comment by fwyzard** [comment]
SLC7/CC7 should be fine

**Comment by smuzaffar** [comment]
but it will fail where vtune is not available  and power8 and aarch builds.  
How about we make this make optional to build ( need some trivial work on scam side) or not make part of release by default and those who wants to use it explicit checkout and build?

**Comment by fwyzard** [comment]
On 29 November 2016 at 16:58, Malik Shahzad Muzaffar <
notifications@github.com> wrote:

> but it will fail where vtune is not available and power8 and aarch builds.
>
Yes.


> How about we make this make optional to build ( need some trivial work on
> scam side) or not make part of release by default and those who wants to
> use it explicit checkout and build?
>
Can the BuildFile be told to build (or not) the plugin based on the
architecture ?

.A


-- 
Strategy is a system of expedients.

Generalfeldmarschall Helmuth Karl Bernhard Graf von Moltke

**Comment by fwyzard** [comment]
@smuzaffar would it help to package the `ittnotify.h` and `libittnotify.a` as a separate external ?
Intel provides them under a dual BSD/GPLv2 license, so we can redistribute them:
```
Instrumentation and Tracing Technology (ITT) API headers and libraries
_____________________________________________________________________________________
List of files:
       include/*
       lib32/libittnotify.*
       lib32/jitprofiling.*
       lib64/libittnotify.*
       lib64/jitprofiling.*

==============================================================================

  The ITT API files are provided under a dual BSD/GPLv2 license.
  When using or redistributing the files, you may do so under either license.
```

**Comment by fwyzard** [comment]
see `/afs/cern.ch/sw/IntelSoftware/linux/x86_64/xe2017/vtune_amplifier_xe_2017.0.2.478468/documentation/en/third-party-programs.txt`

**Comment by smuzaffar** [comment]
yes we can. May be better if we build it from https://github.com/01org/IntelSEAPI/tree/master/ittnotify .... right?

**Comment by fwyzard** [comment]
Sure, I hadn't found that.

**Comment by smuzaffar** [comment]
@iahmad-khan , can you try to build  https://github.com/01org/IntelSEAPI/tree/master/ittnotify 
give it a name ittnotify.

File: FWCore/IntelServices/plugins/BuildFile.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+<use   name="DataFormats/Common"/>
+<use   name="DataFormats/Streamer"/>
+<use   name="FWCore/MessageLogger"/>
+<use   name="FWCore/ParameterSet"/>
+<use   name="FWCore/PluginManager"/>
+<use   name="FWCore/ServiceRegistry"/>
+<use   name="FWCore/Utilities"/>
+<use   name="FWCore/Concurrency"/>
+<use   name="FWCore/Framework"/>
+<use   name="intel-vtune"/>
```
Comment (smuzaffar) [inline FWCore/IntelServices/plugins/BuildFile.xml]: @fwyzard , please replave ```intel-vtune``` with ```ittnotify``` which is now available in 90X IBs

**Comment by fwyzard** [comment]
please test

**Comment by Dr15Jones** [comment]
I don't think the `(pre|post)(Stream|Global){transition name}` callbacks will work correctly since the `pre` and `post` are not guaranteed to happen on the same thread. Also, wouldn't all of these really be _events_ to ITT? That is what you do for `(pre|post)Event`.

**Comment by fwyzard** [comment]
@Dr15Jones , ok, I guess I need a refresher on the thread model used by the framework.
It does make sense that the `Begin(Run|Lumi)` and `End(Run|Lumi)` transitions be on different threads; but I assumed that for a given transition, `preBeginRun` and `postBeginRun` would be called on the same thread ?

Anyway, talking with the Intel people, they were not very clear on the distinction between Events and Tasks: neither allows grouping repetitive tasks, and both allow to select a well-defined time frame.

At the API level, the main difference is that the strings used for task names are kept in a table, while for events they are passed every time. I will have to play with the API a bit to check how it behaves with respect to different threads.

**Comment by Dr15Jones** [comment]
When we start to use multiple threads to process modules during a Run transition, the `postBeginRun` (and `postEndRun`) will trigger based on the last EndPath finishing which could happen on any thread.

**Comment by fwyzard** [comment]
OK, you are saying that there will be the possibility of having
  - thread 1 call `preBeginRun`
  - threads 1, 2, ..., n run the `beginRun` method of different modules in parallel
  - thread x ≠ 1 call `postBeginRun`, depending on the last module to complete the `beginRun` method
?

**Comment by Dr15Jones** [comment]
Yes.

**Comment by davidlange6** [comment]
@fwyzard @Dr15Jones - are there plans for this PR?

**Comment by fwyzard** [comment]
Yes - but I'm still learning the details of the framework "signals" trying
to take into account unscheduled execution and multiple threads... I'll get
it done eventually.

.A

**Pull request closed without merging**